### PR TITLE
chore: deprecate sodiumoxide

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
Changes introduced in this pull request:
- Replace deprecated crate `sodiumoxide` with `argon2` and `xsalsa20poly1305` in `key_management`
- As a side effect, it removed `libsodium-sys` dependency whose build time ranks no.1 in the report generated by `cargo build --timings`
- Fix flacky tests

**Reference issue to close (if applicable)**

Closes https://github.com/ChainSafe/forest/issues/2053

**Other information and links**

### CODE REVIEW

**Comment by lemmih** [comment]
Test suite failure unrelated to this PR. It has been addressed in #2082.

File: key_management/Cargo.toml [hunk 1]
```diff
@@ -20,11 +22,13 @@ rand.workspace               = true
 serde                        = { workspace = true, features = ["derive"] }
 serde_ipld_dagcbor.workspace = true
 serde_json.workspace         = true
-sodiumoxide                  = "0.2"
 thiserror.workspace          = true
+xsalsa20poly1305             = "0.9"
+
```
Comment (lemmih) [inline key_management/Cargo.toml:27]:
```suggestion
```

File: key_management/src/keystore.rs [hunk 2]
```diff
@@ -412,50 +409,66 @@ impl KeyStore {
...
-    fn encrypt(encryption_key: Arc<secretbox::Key>, msg: &[u8]) -> Vec<u8> {
-        let nonce = secretbox::gen_nonce();
-
-        let mut ciphertext = secretbox::seal(msg, &nonce, &encryption_key);
-        ciphertext.append(&mut nonce.as_ref().to_vec());
-        ciphertext
+        let hasher = {
+            let mut param_builder = ParamsBuilder::new();
+            // #define crypto_pwhash_argon2id_MEMLIMIT_INTERACTIVE 67108864U
+            param_builder.m_cost(67108864 / 1024).unwrap();
```
Comment (lemmih) [inline key_management/src/keystore.rs:426]: Can we use `?` here? Is `argon2::Error` well-behaved?

File: key_management/src/keystore.rs [hunk 3]
```diff
@@ -412,50 +409,66 @@ impl KeyStore {
...
+            // KEYBYTES = 0x20
+            // param_builder.output_len(32)?;
+            Argon2::new(
+                argon2::Algorithm::Argon2id,
+                argon2::Version::V0x13,
+                param_builder.params().unwrap(),
+            )
+        };
+        let salt_string =
+            SaltString::b64_encode(&salt).map_err(|err| anyhow::Error::msg(err.to_string()))?;
```
Comment (lemmih) [inline key_management/src/keystore.rs:439]: The salt string errors can't be converted automatically?

**Review by lemmih — APPROVED** [review]
Looks good. I especially like `test_read_old_encrypted_keystore`.

File: key_management/src/keystore.rs (hunk 3 above)
Comment (hanabi1224) [reply key_management/src/keystore.rs:439]: No : )

File: key_management/src/keystore.rs (hunk 2 above)
Comment (hanabi1224) [reply key_management/src/keystore.rs]: Fixed. Need to move it out and map_err tho

File: key_management/Cargo.toml (hunk 1 above)
Comment (hanabi1224) [reply key_management/Cargo.toml:27]: fixed

File: Cargo.lock [hunk 4]
```diff
@@ -5563,18 +5594,6 @@ dependencies = [
  "libsecp256k1-core",
 ]
 
-[[package]]
-name = "libsodium-sys"
```
Comment (hanabi1224) [inline Cargo.lock:5567]: It's gone

File: key_management/src/keystore.rs [hunk 5]
```diff
@@ -412,59 +410,77 @@ impl KeyStore {
...
+        let mut param_builder = ParamsBuilder::new();
+        // #define crypto_pwhash_argon2id_MEMLIMIT_INTERACTIVE 67108864U
+        param_builder
+            .m_cost(67108864 / 1024)
+            .map_err(map_err_to_anyhow)?;
+        // #define crypto_pwhash_argon2id_OPSLIMIT_INTERACTIVE 2U
+        param_builder.t_cost(2).map_err(map_err_to_anyhow)?;
+        // https://docs.rs/sodiumoxide/latest/sodiumoxide/crypto/secretbox/xsalsa20poly1305/constant.KEYBYTES.html
+        // KEYBYTES = 0x20
+        // param_builder.output_len(32)?;
```
Comment (LesnyRumcajs) [inline key_management/src/keystore.rs:433]: Do we need these leftover comments? I mean, it's helpful to know where did `67108864U` come from; perhaps a link to the C/C++ implementation where it's defined would be more beneficial for the reader.

**Review by LesnyRumcajs — APPROVED** [review]
Nice, especially given that those libraries are far from being ergonomic. But we don't have anything better in the ecosystem right now if I understand correctly.

File: key_management/src/keystore.rs [hunk 6]
```diff
@@ -412,59 +410,77 @@ impl KeyStore {
 impl EncryptedKeyStore {
     fn derive_key(
         passphrase: &str,
-        prev_salt: Option<Vec<u8>>,
-    ) -> Result<(pwhash::Salt, Arc<secretbox::Key>), EncryptedKeyStoreError> {
+        prev_salt: Option<[u8; RECOMMENDED_SALT_LEN]>,
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply key_management/src/keystore.rs]: fixed

File: key_management/src/keystore.rs (hunk 5 above)
Comment (hanabi1224) [reply key_management/src/keystore.rs]: links added

**Pull request merged**